core/src/main/scala/org/apache/spark/h2o/converters/PrimitiveRDDConverter.scala

```diff
@@ -57,25 +65,15 @@ private[converters] object PrimitiveRDDConverter extends Logging with ConverterU
   def perPrimitiveRDDPartition[T]() // extra arguments for this transformation
                                  (keyName: String, vecTypes: Array[Byte], uploadPlan: Option[immutable.Map[Int, NodeDesc]]) // general arguments
                                  (context: TaskContext, it: Iterator[T]): (Int, Long) = { // arguments and return types needed for spark's runJob input
-    val con = ConverterUtils.getWriteConverterContext(uploadPlan, context.partitionId())
 
+    val (iterator, dataSize) = WriteConverterCtxUtils.bufferedIteratorWithSize(uploadPlan, it)
+    val con = WriteConverterCtxUtils.create(uploadPlan, context.partitionId(), dataSize)
     con.createChunks(keyName, vecTypes, context.partitionId())
-    // try to wait for reply to ensure we can continue with sending
-    it.foreach { r =>
-      r match {
-        case n: Number => con.put(0, n)
-        case n: Boolean => con.put(0, n)
-        case n: String => con.put(0, n)
-        case n: java.sql.Timestamp => con.put(0, n)
-        case _ => con.putNA(0)
-      }
-      con.increaseRowCounter()
-    }
+    iterator.foreach {con.putAnySupportedType(0, _)}
     //Compress & write data in partitions to H2O Chunks
     con.closeChunks()
-
     // Return Partition number and number of rows in this partition
-    (context.partitionId, con.numOfRows)
+    (context.partitionId, iterator.size)
```

Comment:
I am just saying, that we are breaking iterator invariant by knowing internal implementation If somebody changes internal implementation, we can be affected here. From my point of view, it is safer to pass here number of invocations from foreach above.